encode_utils/connection.py

```diff
@@ -1227,6 +1227,131 @@ def remove_props(self, rec_id, props=[]):
+                ``requests.exceptions.HTTPError`` if a 403 status (forbidden)
+                is returned. If set to `False` and there still is a 403 return
+                status, then the object you were trying to PATCH will be
+                fetched from the Portal in JSON format as this function's
+                return value.
+            extend_array_values: `bool`. Only affects keys with array values.
+                `True` (default) means to extend the corresponding value on the
+                Portal with what's specified in the payload. `False` means to
+                replace the value on the Portal with what's in the payload.
+
+        Returns:
+            `dict`: The JSON response from the PUT operation, or an empty dict
+                if the record doesn't exist on the Portal. Will also be an
+                empty dict if the dry-run feature is enabled.
+
+        Raises:
+            requests.exceptions.HTTPError: if the return status is not ok
+                (excluding a 403 status if 'raise_403' is False).
+        """
+        # Make sure we have a payload that can be converted to valid JSON, and
+        # tuples become arrays, ...
+        patch = json.loads(json.dumps(patch))
+        self.debug_logger.debug("\nIN remove_and_patch()")
+        encode_id = patch[self.ENCID_KEY]
+        rec_json = self.get(rec_ids=encode_id, frame="object", ignore404=True)
+        if not rec_json:  # Ensure that the record exists on the Portal:
+            return {}
+        profile = eup.Profile(rec_json["@id"])
+        payload = self.get(rec_ids=encode_id, frame="edit", ignore404=False)
+        for prop in props:
+            if profile.is_prop_required(prop):
+                raise Exception("Can't remove required property")
+            elif profile.is_prop_not_submittable(prop):
+                raise Exception("Can't remove non-submittable property.")
+            elif profile.is_prop_read_only(prop):
+                raise Exception("Can't remove read-only property.")
+            else:
+                # Then it is safe to remove this property.
+                payload.pop(prop, None)
+        for key in patch:
+            if (
+                extend_array_values
+                and isinstance(patch[key], list)
+                and key in payload
+            ):
+                val = payload[key]
+                val.extend(patch[key])
+                # CHECK FOR DUPLICATES: Be careful as some can be tricky, i.e.
+                # ['/documents/id1', 'id1'] such a duplicate should be
+                # identified and removed, leaving us with ["id1"].
+                # Checks for arrays of strings or of dicts.
+                if isinstance(val[0], str):
+                    payload[key] = eup.remove_duplicate_associations(val)
+                elif isinstance(val[0], dict):
+                    payload[key] = eup.remove_duplicate_objects(val)
+            else:
+                payload[key] = patch[key]
```

Comment:
Do we want to do anything special when the user tries to patch a property that was in the props to remove also? I'm fine with the current behavior since the docstring makes it clear that the props are removed before patching, but maybe you have a different opinion. If it's not OK, then we need to check for it and then raise an error.

Reply:
I think I'm OK with the current behavior.